BootloaderCommonPkg/Library/ShellLib/CmdCorruptComponent.c

```diff
@@ -0,0 +1,201 @@
+  UINT32              BiosStart;
+  UINT8               *Buf;
+  UINT32              CompOffset;
+  UINT32              CompSize;
+  UINT32              RandOffset;
+  UINT32              ChunkAddr;
+  UINT32              ChunkAddrAligned;
+  UINT8*              CurrVal;
+  UINT16              RandVal;
+  UINT8               NewVal;
+
+  Status = EFI_SUCCESS;
+
+  if (Argc == 3) {
+    switch (StrDecimalToUintn (Argv[1])) {
+      case 0:
+        IsBackup = FALSE;
+        break;
+      case 1:
+        IsBackup = TRUE;
+        break;
+      default:
+        ShellPrint (L"Error, invalid boot partition\n");
+        ShowUsage ();
+        return EFI_INVALID_PARAMETER;
+    }
+    CompStr = Argv[2];
+    if (StrLen(CompStr) != 4) {
+      ShellPrint (L"Error, invalid component\n");
+      ShowUsage ();
+      return EFI_INVALID_PARAMETER;
+    }
+    CompSig = SIGNATURE_32 (CompStr[0], CompStr[1], CompStr[2], CompStr[3]);
+  } else if (Argc < 3) {
+    ShellPrint (L"Error, too few arguments\n");
+    ShowUsage ();
+    return EFI_INVALID_PARAMETER;
+  } else if (Argc > 3) {
+    ShellPrint (L"Error, too many arguments\n");
+    ShowUsage ();
+    return EFI_INVALID_PARAMETER;
+  }
+
+  SpiService = (SPI_FLASH_SERVICE *) GetServiceBySignature (SPI_FLASH_SERVICE_SIGNATURE);
+  if (SpiService == NULL) {
+    ShellPrint (L"Error occured getting SPI flash service");
+    return EFI_UNSUPPORTED;
+  }
+
+  Status = SpiService->SpiInit ();
+  if (EFI_ERROR(Status)) {
+    ShellPrint (L"Error occured initializing SPI flash service %r\n", Status);
+    return Status;
+  }
+
+  Status = SpiService->SpiGetRegion (FlashRegionBios, NULL, &BiosStart);
+  if (EFI_ERROR(Status)) {
+    ShellPrint (L"Error occured getting BIOS SPI flash address %r\n", Status);
+    return Status;
+  }
+
+  Buf = AllocatePages (EFI_SIZE_TO_PAGES (SIZE_4KB));
+  if (Buf == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Status = GetComponentInfoByPartition (CompSig, IsBackup, &CompOffset, &CompSize);
+  if (EFI_ERROR (Status)) {
+    ShellPrint (L"Error, component info unable to be retrieved\n");
+    return Status;
+  }
+
+  // Choose a random offset in the component to corrupt
+  if (!GetRandomNumber32 (&RandOffset)) {
+    ShellPrint (L"Error, random offset unable to be generated\n");
+    return EFI_UNSUPPORTED;
+  }
+
+  RandOffset %= CompSize;
+  ChunkAddr = BiosStart + CompOffset + RandOffset;
+  ChunkAddrAligned = ALIGN_DOWN (ChunkAddr, SIZE_4KB);
+
+  // Read the whole 4KB chunk containing the offset from SPI flash
+  Status = SpiService->SpiRead (FlashRegionBios,
+                                ChunkAddrAligned,
+                                SIZE_4KB,
+                                Buf);
+  if (EFI_ERROR(Status)) {
+    ShellPrint (L"Error occured in SPI flash read %r\n", Status);
+    return Status;
+  }
+
+  // Erase the whole 4KB chunk containing the offset from SPI flash
+  Status = SpiService->SpiErase (FlashRegionBios, ChunkAddrAligned, SIZE_4KB);
```

Comment:
Is the complete 4K block is over written?

Reply:
We read out the entire 4KB block, but only really change 1B of it. Since SPI flash writes are bitwise-AND operations and since we want to test scenarios where 0's are inverted to 1's, we need to execute a SPI flash erase beforehand. SPI flash erase operations are only allowed to be executed in 4KB blocks.